File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
// Depth-First-Search
// Time and Space: O(V + E)
var validPath = function(n, edges, source, destination) {
    if (source === destination) return true;
    const adjList = new Map(), visited = new Set(), stack = []

    for (let [x , y] of edges) {
        if (!adjList.has(x)) adjList.set(x, new Set());
        if (!adjList.has(y)) adjList.set(y, new Set());
        adjList.get(x).add(y);
        adjList.get(y).add(x);
    }

    if (adjList.has(source)) stack.push(source)

    while (stack.length) {
        const nde = stack.pop();
        visited.add(nde);
        if (nde === destination) return true;
        if (adjList.has(nde)) {
            const nodes = adjList.get(nde)
            if (nodes.has(destination)) return true;
            Array.from(nodes).forEach(n => {
                if (!visited.has(n)) {
                    visited.add(n)
                    stack.push(n);
                }
            })
        }
    }
    return false;
};

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.swift
class Solution {
    func validPath(_ n: Int, _ edges: [[Int]], _ source: Int, _ destination: Int) -> Bool {
        var adjList: [Int: Set<Int>] = [:]

        for edge in edges {
            if adjList[edge[0]] == nil { adjList[edge[0]] = Set<Int>()}
            if adjList[edge[1]] == nil { adjList[edge[1]] = Set<Int>()}
            adjList[edge[0]]!.insert(edge[1])
            adjList[edge[1]]!.insert(edge[0])
        }
        var stack: [Int] = [source], visited: Set<Int> = Set<Int>()
        while !stack.isEmpty {
            var node: Int = stack.removeFirst()
            if node == destination { return true }
            visited.insert(node)
            if var nodes: Set<Int> = adjList[node] {
                if nodes.contains(destination) { return true }
                for n in nodes {
                    if !visited.contains(n) {
                        stack.append(n)
                        visited.insert(n)
                    }
                }
            }
        }
        return false
    }
}